File: app/src/main/java/dev/blackcat/minauta/MainActivity.java
package dev.blackcat.minauta;

import android.content.Intent;
import android.os.Bundle;
import android.view.View;
import android.widget.Button;
import android.widget.TextView;

import java.util.Calendar;

import dev.blackcat.minauta.data.Account;
import dev.blackcat.minauta.net.Connection;
import dev.blackcat.minauta.net.ConnectionFactory;
import dev.blackcat.minauta.net.FakeConnection;
import dev.blackcat.minauta.data.store.PreferencesStore;

public class MainActivity extends MyAppCompatActivity
{

    Button configureButton;
    Button startButton;
    TextView accountTextView;

    @Override
    protected void onCreate(Bundle savedInstanceState)
    {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        this.initComponents();
    }

    @Override
    protected void onResume()
    {
        super.onResume();

        Account account = new PreferencesStore(this).getAccount();
        if (account == null)
        {
            accountTextView.setText(R.string.configure_account_text);
            startButton.setEnabled(false);
        }
        else
        {
            if (account.getSession() == null)
            {
                accountTextView.setText(account.getUsername());
                startButton.setEnabled(true);
            }
            else
            {
                Intent intent = new Intent(MainActivity.this, SessionActivity.class);
                startActivity(intent);
            }
        }
    }

    private void initComponents()
    {
        accountTextView = this.getViewById(R.id.accountTextView);

        configureButton = this.getViewById(R.id.configureButton);
        configureButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view)
            {
                Intent intent = new Intent(MainActivity.this, AccountActivity.class);
                startActivity(intent);
            }
        });

        startButton = this.getViewById(R.id.startButton);
        startButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view)
            {
                startSession();
            }
        });
    }

    private void startSession()
    {
        PreferencesStore store = new PreferencesStore(this);
        Account account = store.getAccount();

        // TODO: Change for a real connection
        Connection connection = ConnectionFactory.createSessionProducer(FakeConnection.class);
        Connection.LoginResult result = connection.login(account);

        if (result.state != Connection.State.OK)
        {
            int resId = R.string.login_error;
            switch (result.state)
            {
                case ALREADY_CONNECTED:
                    resId = R.string.already_connected_error;
                    break;
                case INCORRECT_PASSWORD:
                    resId = R.string.incorrect_password_error;
                    break;
                case UNKNOWN_USERNAME:
                    resId = R.string.unknown_username_error;
                    break;
            }
            this.showDialogWithText(resId);
            return;
        }

        long now = Calendar.getInstance().getTimeInMillis();
        store.setSession(result.session.getLogoutParams(), result.session.getTimeParams(), now);
        Intent intent = new Intent(this, SessionActivity.class);
        startActivity(intent);
    }
}

File: app/src/main/java/dev/blackcat/minauta/ui/main/MainActivity.kt
package dev.blackcat.minauta.ui.main

import android.os.Bundle
import android.text.Editable
import android.view.View
import android.widget.*
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import dev.blackcat.minauta.R
import dev.blackcat.minauta.data.Account
import dev.blackcat.minauta.data.SessionLimit
import dev.blackcat.minauta.data.SessionTimeUnit
import dev.blackcat.minauta.ui.MyAppCompatActivity

class MainActivity : MyAppCompatActivity() {

    lateinit var viewModel: MainViewModel

    lateinit var portalButton: Button
    lateinit var configureButton: Button
    lateinit var startButton: Button
    lateinit var accountTextView: TextView

    lateinit var sessionLimitCheckBox: CheckBox
    lateinit var sessionLimitEditText: EditText
    lateinit var sessionLimitSpinner: Spinner

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        viewModel = ViewModelProvider.AndroidViewModelFactory(application).create(MainViewModel::class.java)

        findViewById<ImageView>(R.id.logoImageView).setOnClickListener { showAboutDialog() }
        findViewById<TextView>(R.id.appNameTextView).setOnClickListener { showAboutDialog() }

        accountTextView = findViewById(R.id.accountTextView)

        configureButton = findViewById(R.id.configureButton)
        configureButton.setOnClickListener {
            viewModel.startAccountActivity(this)
        }

        portalButton = findViewById(R.id.portalButton)
        portalButton.setOnClickListener {
            viewModel.startPortalActivity(this)
        }

        sessionLimitEditText = findViewById(R.id.sessionLimitEditText)
        sessionLimitEditText.text = Editable.Factory.getInstance().newEditable("0")
        sessionLimitEditText.visibility = View.GONE

        sessionLimitSpinner = findViewById(R.id.sessionLimitSpinner)
        sessionLimitSpinner.visibility = View.GONE

        sessionLimitCheckBox = findViewById(R.id.sessionLimitCheckBox)
        sessionLimitCheckBox.setOnCheckedChangeListener { buttonView, isChecked ->
            val visibility = if (isChecked) View.VISIBLE else View.GONE
            sessionLimitEditText.visibility = visibility
            sessionLimitSpinner.visibility = visibility
        }

        startButton = findViewById(R.id.startButton)
        startButton.setOnClickListener {
            val sessionLimit = createSessionLimit()
            viewModel.startSession(this, sessionLimit)
        }

        viewModel.accountObservable.observe(this, Observer { account ->
            checkAccountState(account)
            setSessionLimit(account.sessionLimit)
        })

        viewModel.sessionObservable.observe(this, Observer { session ->
            session?.let {
                viewModel.startSessionActivity(this)
            }
        })
    }

    override fun onResume() {
        super.onResume()
        viewModel.checkAccount()
        viewModel.checkSession()
    }

    override fun onPause() {
        super.onPause()
        val sessionLimit = createSessionLimit()
        viewModel.saveSessionLimit(sessionLimit)
    }

    private fun checkAccountState(account: Account) {
        when (account.initialized) {
            false -> {
                accountTextView.setText(R.string.configure_account_text)
                startButton.isEnabled = false
                portalButton.isEnabled = false
            }
            true -> {
                accountTextView.text = account.username
                startButton.isEnabled = true
                portalButton.isEnabled = true
            }
        }
    }

    private fun setSessionLimit(sessionLimit: SessionLimit) {
        sessionLimitCheckBox.isChecked = sessionLimit.enabled
        sessionLimitEditText.text = Editable.Factory.getInstance().newEditable("${sessionLimit.time}")
        sessionLimitSpinner.setSelection(sessionLimit.timeUnit.value)
    }

    private fun createSessionLimit(): SessionLimit {
        val unit = SessionTimeUnit.fromInt(sessionLimitSpinner.selectedItemPosition)
        val time = sessionLimitEditText.text.toString().toInt()
        return SessionLimit(sessionLimitCheckBox.isChecked, time, unit)
    }

}
